### PR TITLE
citgm: add a timer to time each module test

### PR DESCRIPTION
**Pull request by GeorgeAdams95**
This will be useful for diagnosing machine issues and will also allow us to see which modules are slowing down the citgm-all job.

### CODE REVIEW

**Comment by gibfahn** [comment]
If we're going to do this, we should put it in the `duration_ms` field of the tap/Junit output so we can see it in Jenkins.

**Comment by Fishrock123** [comment]
This sounds useful to me. @gibfahn makes a good point too.

**Comment by mhdawson** [comment]
+1 from me  as well.

**Comment by GeorgeAdams95** [comment]
@TheAlphaNerd do you want the tap output to always show the timer?

**Comment by gibfahn** [comment]
@GeorgeAdams95 Assuming you mean showing the duration even if the test passes and there is no output, then I'd say yes.

**Comment by TheAlphaNerd** [comment]
+1 to what Gibson said

**Comment by GeorgeAdams95** [comment]
@TheAlphaNerd should be good to merge? any changes you want me to make

File: lib/reporter/tap.js [hunk 1]
```diff
@@ -11,12 +11,10 @@ function generateTest(mod, count) {
...
     mod.error = mod.error.message;
   }
-  var output = '';
   var error = mod.error ? [directive, mod.error].join(' ') : '';
-  if (mod.testOutput && mod.testOutput.length > 0) {
-    output = '\n  ---\n' + util.sanitizeOutput(mod.testOutput, '  #') + '\n  ...';
-  }
-  return [result, count + 1, '-', mod.name, 'v' + mod.version + error + output].join(' ');
+  var duration = '\n  duration_ms: ' + mod.duration;
+  var output = mod.testOutput ? '\n' + util.sanitizeOutput(mod.testOutput, '  #') : '';
```
Comment (TheAlphaNerd) [inline lib/reporter/tap.js]: why did you remove the test to ensure that `mod.testOutput` exists?

Comment (TheAlphaNerd) [inline lib/reporter/tap.js]: oh nvm, I see the ternary

**Comment by GeorgeAdams95** [comment]
@TheAlphaNerd I'll update the json to include mod.duration

**Comment by GeorgeAdams95** [comment]
Updated and rebased PTAL

File: bin/citgm.js [hunk 2]
```diff
@@ -84,8 +85,10 @@ function launch(mod, options) {
   }).on('data', function(type, key, message) {
     log[type](key, message);
   }).on('end', function(module) {
+    module.duration = new Date() - start;
     reporter.logger(log, module);
 
+    log.info('duration', 'test time: ' + module.duration + 'ms');
```
Comment (TheAlphaNerd) [inline bin/citgm.js]: nit: this should be `test duration:`

will fix this on landing if it is the only issue

**Comment by TheAlphaNerd** [comment]
Going to do a run of citgm with this PR and see what the output looks like.

If everything is the way we want it to be I'll land this

https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/459/

**Comment by GeorgeAdams95** [comment]
@TheAlphaNerd fixed that nit

**Comment by GeorgeAdams95** [comment]
duration is returning as nan: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/459/

**Comment by GeorgeAdams95** [comment]
@TheAlphaNerd fixed, please re-run on ci.nodejs.org when you get a chance

**Comment by TheAlphaNerd** [comment]
what was the fix? did you include a new test?

**Comment by GeorgeAdams95** [comment]
I had stupidly declared the start time in the wrong place :1234:

**Comment by TheAlphaNerd** [comment]
is there any way that we can write a test to ensure this doesn't break in
the future?

**Comment by GeorgeAdams95** [comment]
@TheAlphaNerd not that I can think of, unless we can run citgm-all, check that the output contains a duration and check that it doesn't equal NaN

**Comment by GeorgeAdams95** [comment]
should we be outputting the duration in our xml results as well or just tap?

**Comment by TheAlphaNerd** [comment]
it should be in all reporters

**Comment by GeorgeAdams95** [comment]
added junit support

**Comment by GeorgeAdams95** [comment]
Should I add a junit xml parser to one of the tests to match the tap parser that we now use? This would allow us to ensure that we are reporting the duration correctly

**Comment by TheAlphaNerd** [comment]
that's not a bad idea. I'd like to see this information in the markdown
reporter too

**Comment by GeorgeAdams95** [comment]
I have now added a junit xml parser. How would I go about adding the duration to the markdown reporter? there seems to be no obvious existing specification to use

**Comment by GeorgeAdams95** [comment]
I have now added duration to the markdown logger PTAL

**Comment by GeorgeAdams95** [comment]
@TheAlphaNerd here is the re-run of the jenkins job showing duration output: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/460/

**Comment by GeorgeAdams95** [comment]
@TheAlphaNerd is this good to merge now?

File: lib/reporter/junit.js [hunk 3]
```diff
@@ -11,6 +11,7 @@ function generateTest(xml, mod) {
   var output =util.sanitizeOutput(mod.testOutput, '', true);
   var item = xml.ele('testcase');
   item.att('name', [mod.name, 'v' + mod.version].join('-'));
+  item.att('time', mod.duration/1000);
```
Comment (TheAlphaNerd) [inline lib/reporter/junit.js]: nit: there should be spaces between mod.duration `/` and 1000

File: lib/reporter/markdown.js [hunk 4]
```diff
@@ -8,7 +8,7 @@ function printModulesMarkdown(logger, title, modules) {
   if (modules.length > 0) {
     logger('### ' + title + ' Modules');
     _.each(modules, function (mod) {
-      logger('  * ' + mod.name + ' v' + mod.version);
+      logger('  * ' + mod.name + ' v' + mod.version + ' ' + mod.duration + 'ms');
```
Comment (TheAlphaNerd) [inline lib/reporter/markdown.js]: perhaps include the word `duration` in the output?

File: test/fixtures/parsed-junit.json [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+      "name": "citgm"
+    },
+    "testcase": [
+      {
+        "$": {
+          "name": "iPass-v4.2.2",
+          "time": "0.05"
+        },
+        "system-out": [
+          "\n      \n    "
```
Comment (TheAlphaNerd) [inline test/fixtures/parsed-junit.json]: are you able to replace the spaces here with `\t`? Totally cool if not

File: test/fixtures/parsed-junit.json [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        ]
+      },
+      {
+        "$": {
+          "name": "iFlakyFail-v3.3.3",
+          "time": "0.05"
+        },
+        "system-out": [
+          "\n       Thanks for testing!\n    "
```
Comment (TheAlphaNerd) [inline test/fixtures/parsed-junit.json]: same as above

File: test/fixtures/parsed-junit.json [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+          }
+        ]
+      },
+      {
+        "$": {
+          "name": "iFail-v3.0.1",
+          "time": "0.05"                                          
+        },
+        "system-out": [
+          "\n       Thanks for testing!\n    "
```
Comment (TheAlphaNerd) [inline test/fixtures/parsed-junit.json]: same as above

File: test/fixtures/parsed-junit.json (hunk 5 above)
Comment (GeorgeAdams95) [inline test/fixtures/parsed-junit.json]: No, this does not work

**Comment by GeorgeAdams95** [comment]
@TheAlphaNerd  Updated PTAL

**Comment by TheAlphaNerd** [comment]
landed in 7c77d61637
